File: pure_cnn_react/src/model/conv_layer.js
import Layer from './layer'
import Mat from '../util/mat'
import Shape from '../util/mat_shape'

import {
    INIT_ZEROS,
    INIT_RANDN,
    LAYER_CONV,
    ACTIVATION_RELU
} from '../util/constant';

export default class ConvLayer extends Layer {
    constructor(name, units, kernel_width, kernel_height, kernel_stride_x, kernel_stride_y, padding) {
        super(name, units);

        this.type = LAYER_CONV;
        this.activation = ACTIVATION_RELU;

        this.kernel_width = kernel_width;
        this.kernel_height = kernel_height;
        this.kernel_depth = undefined;

        this.kernel_stride_x = 1;
        this.kernel_stride_y = 1;
        this.pad_x = 0;
        this.pad_y = 0;

        this.kernel = new Array(units);
        this.biases = new Float32Array(units);
        this.kernel_grad = new Array(this.units);
        this.biases_grad = new Float32Array(this.units);

        this.back_error = undefined;
    }

    set_input_layer(inputLayer) {
        super.set_input_layer(inputLayer);

        const output_width = Math.floor((this.input_shape.width + this.pad_x * 2 - this.kernel_width) / this.kernel_stride_x + 1);
        const output_height = Math.floor((this.input_shape.height + this.pad_y * 2 - this.kernel_height) / this.kernel_stride_y + 1);

        this.output_shape = new Shape(output_width, output_height, this.units);
        this.kernel_depth = this.input_shape.depth;

        const kernel_shape = new Shape(this.kernel_width, this.kernel_height, this.kernel_depth);

        for (let j = 0; j < this.units; j++) {
            this.kernel[j] = new Mat(kernel_shape, INIT_RANDN);
            this.kernel_grad[j] = new Mat(kernel_shape, INIT_ZEROS);
        }
    }

    set_params(weight, bias) {
        for (let i = 0; i < this.units; i++) {
            this.kernel[i].set_value(weight[i]);
        }
        this.biases = new Float32Array(bias);
    }

    forward() {
        this.input = this.pre_layer.output;
        this.output = new Array(this.network.mini_batch_size);

        for (let i = 0; i < this.network.mini_batch_size; i++) {
            this.output[i] = new Mat(this.output_shape, INIT_ZEROS);

            for (let j = 0; j < this.units; j++) {
                for (let out_y = 0; out_y < this.output_shape.height; out_y++) {
                    for (let out_x = 0; out_x < this.output_shape.width; out_x++) {
                        for (let ker_y = 0; ker_y < this.kernel_height; ker_y++) {
                            let input_y = out_y + ker_y;
                            for (let ker_x = 0; ker_x < this.kernel_width; ker_x++) {
                                let input_x = out_x + ker_x;
                                for (let ker_d = 0; ker_d < this.kernel_depth; ker_d++) {
                                    this.output[i].add_value_by_coordinate(out_x, out_y, j, this.kernel[j].get_value_by_coordinate(ker_x, ker_y, ker_d) * this.input[i].get_value_by_coordinate(input_x, input_y, ker_d));
                                }
                            }
                        }
                        this.output[i].add_value_by_coordinate(out_x, out_y, j, this.biases[j]);
                    }
                }
            }
            this.output[i].activate(this.activation);
        }
    }

    backward() {
        let next_layer = this.next_layer;

        this.back_error = new Array(this.network.mini_batch_size);

        for (let i = 0; i < this.network.mini_batch_size; i++) {
            this.back_error[i] = new Mat(this.input_shape, INIT_ZEROS);

        }

        for (let i = 0; i < this.network.mini_batch_size; i++) {
            for (let j = 0; j < this.units; j++) {
                for (let out_y = 0; out_y < this.output_shape.height; out_y++) {
                    for (let out_x = 0; out_x < this.output_shape.width; out_x++) {
                        let error_delta = next_layer.back_error[i].get_value_by_coordinate(out_x, out_y, j) * ( ( this.output[i].get_value_by_coordinate(out_x, out_y, j) > 0 ) ? 1 : 0 );
                        let error_delta_with_learning_rate = -this.network.batch_learning_rate * error_delta;
                        this.biases_grad[j] += error_delta_with_learning_rate;
                        for (let ker_d = 0; ker_d < this.kernel_depth; ++ker_d) {
                            for (let ker_y = 0; ker_y < this.kernel_height; ++ker_y) {
                                let input_y = ker_y + out_y;
                                for (let ker_x = 0; ker_x < this.kernel_width; ++ker_x) {
                                    let input_x = ker_x + out_x;
                                    this.kernel_grad[j].add_value_by_coordinate(ker_x, ker_y, ker_d, error_delta_with_learning_rate * this.input[i].get_value_by_coordinate(input_x, input_y, ker_d));
                                    this.back_error[i].add_value_by_coordinate(input_x, input_y, ker_d, error_delta * this.kernel[j].get_value_by_coordinate(ker_x, ker_y, ker_d));
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    batch_update() {
        const l2_regularization = 1.0 - this.network.batch_learning_rate * this.network.l2;

        for (let j = 0; j < this.units; j++) {
            this.kernel[j].operation_scale_and_add_mat(l2_regularization, this.kernel_grad[j]);
            this.biases[j] += this.biases_grad[j];

            this.kernel_grad[j].operation_scale_mat(this.network.momentum);
            this.biases_grad[j] *= this.network.momentum;
        }
    }
}

File: pure_cnn_py/model/conv_layer.py
from pure_cnn_py.model.layer import Layer
from pure_cnn_py.util.mat import Mat
from pure_cnn_py.util.mat_shape import MatShape
from pure_cnn_py.util.constant import INIT_ZEROS, \
    INIT_RANDN, \
    LAYER_CONV, \
    ACTIVATION_RELU


class ConvLayer(Layer):
    def __init__(self, name, units, kernel_width, kernel_height, kernel_stride_x, kernel_stride_y, padding, network):
        super(ConvLayer, self).__init__(name, units)

        self.layer_type = LAYER_CONV
        self.activation = ACTIVATION_RELU
        self.input_layer = None

        self.kernel_width = kernel_width
        self.kernel_height = kernel_height
        self.kernel_depth = None

        self.kernel_stride_x = 1
        self.kernel_stride_y = 1
        self.pad_x = 0
        self.pad_y = 0

        self.kernel = [0] * self.units
        self.biases = [0] * self.units
        self.kernel_grad = [0] * units
        self.biases_grad = [0] * units

        self.back_error = None

        self.output_shape = None

        self.network = network
        self.input = None
        self.output = None

    def set_input_layer(self, input_layer):
        super(ConvLayer, self).set_input_layer(input_layer)

        output_width = (self.input_shape.width + self.pad_x * 2 - self.kernel_width) // self.kernel_stride_x + 1
        output_height = (self.input_shape.height + self.pad_y * 2 - self.kernel_height) // self.kernel_stride_y + 1

        self.output_shape = MatShape(output_width, output_height, self.units)
        self.kernel_depth = self.input_shape.depth

        kernel_shape = MatShape(self.kernel_width, self.kernel_height, self.kernel_depth)

        for j in range(0, self.units):
            self.kernel[j] = Mat(kernel_shape, INIT_RANDN)
            self.kernel_grad[j] = Mat(kernel_shape, INIT_ZEROS)

    def set_params(self, weight, bias):
        for i in range(0, self.units):
            self.kernel[i].set_value(weight[i])

        self.biases = bias

    def forward(self):
        self.input = self.pre_layer.output
        self.output = [0] * self.network.mini_batch_size

        for i in range(0, self.network.mini_batch_size):
            self.output[i] = Mat(self.output_shape, INIT_ZEROS)

            for j in range(0, self.units):
                for out_y in range(0, self.output_shape.height):
                    for out_x in range(0, self.output_shape.width):
                        for ker_y in range(0, self.kernel_height):
                            input_y = out_y + ker_y
                            for ker_x in range(0, self.kernel_width):
                                input_x = out_x + ker_x
                                for ker_d in range(0, self.kernel_depth):
                                    self.output[i].add_value_by_coordinate(out_x,
                                                                           out_y,
                                                                           j,
                                                                           self.kernel[j].get_value_by_coordinate(ker_x,
                                                                                                                  ker_y,
                                                                                                                  ker_d) *
                                                                           self.input[i].get_value_by_coordinate(
                                                                               input_x, input_y, ker_d))
                        self.output[i].add_value_by_coordinate(out_y, out_x, j, self.biases[j])
            self.output[i].activate(self.activation)

    def backward(self):
        next_layer = self.next_layer

        self.back_error = [0] * self.network.mini_batch_size

        for i in range(0, self.network.mini_batch_size):
            self.back_error[i] = Mat(self.input_shape, INIT_ZEROS)

        for i in range(0, self.network.mini_batch_size):
            for j in range(0, self.units):
                for out_y in range(0, self.output_shape.height):
                    for out_x in range(0, self.output_shape.width):
                        error_delta = next_layer.back_error[i].get_value_by_coordinate(out_x, out_y, j) * (
                            1 if self.output[i].get_value_by_coordinate(out_x, out_y, j) > 0 else 0)
                        error_delta_with_learning_rate = -1 * self.network.batch_learning_rate * error_delta
                        self.biases_grad[j] += error_delta_with_learning_rate

                        for ker_d in range(self.kernel_depth):
                            for ker_y in range(self.kernel_height):
                                input_y = ker_y + out_y
                                for ker_x in range(self.kernel_width):
                                    input_x = ker_x + out_x

                                    self.kernel_grad[j].add_value_by_coordinate(ker_x, ker_y, ker_d,
                                                                                error_delta_with_learning_rate *
                                                                                self.input[i].get_value_by_coordinate(
                                                                                    input_x, input_y, ker_d))

                                    self.back_error[i].add_value_by_coordinate(input_x, input_y, ker_d,
                                                                               error_delta * self.kernel[
                                                                                   j].get_value_by_coordinate(ker_x,
                                                                                                              ker_y,
                                                                                                              ker_d))

    def batch_update(self):
        l2_regularization = 1 - self.network.batch_learning_rate * self.network.l2

        for j in range(0, self.units):
            self.kernel[j].operation_scale_and_add_mat(l2_regularization, self.kernel_grad[j])
            self.biases[j] += self.biases_grad[j]

            self.kernel_grad[j].operation_scale_mat(self.network.momentum)
            self.biases_grad[j] *= self.network.momentum
